pilot/pkg/config/kube/gateway/conversion.go

```diff
@@ -641,161 +727,258 @@ func extractParentReferenceInfo(gateways map[parentKey][]*parentInfo, routeRefs
+					Tls: routes,
+					Tcp: tcpRoutes,
+				},
+			})
 		}
-		routes = append(routes, ir)
 	}
+	return vs
+}
 
-	reportError(nil)
-	gatewayNames := referencesToInternalNames(parentRefs)
-	if len(gatewayNames) == 0 {
-		// TODO we need to properly return not admitted here
-		return nil
+func convertTCPRoute(ctx configContext, r k8s.TCPRouteRule, obj config.Config, enforceRefGrant bool) (*istio.TCPRoute, *ConfigError) {
+	if tcpWeightSum(r.BackendRefs) == 0 {
+		// The spec requires us to reject connections when there are no >0 weight backends
+		// We don't have a great way to do it. TODO: add a fault injection API for TCP?
+		return &istio.TCPRoute{
+			Route: []*istio.RouteDestination{{
+				Destination: &istio.Destination{
+					Host:   "internal.cluster.local",
+					Subset: "zero-weight",
+					Port:   &istio.PortSelector{Number: 65535},
+				},
+				Weight: 0,
+			}},
+		}, nil
 	}
-	configs := make([]config.Config, 0, len(route.Hostnames))
-	for i, host := range hostnameToStringList(route.Hostnames) {
-		name := fmt.Sprintf("%s-tls-%d-%s", obj.Name, i, constants.KubernetesGatewayName)
-		// Create one VS per hostname with a single hostname.
-		// This ensures we can treat each hostname independently, as the spec requires
-		vsConfig := config.Config{
-			Meta: config.Meta{
-				CreationTimestamp: obj.CreationTimestamp,
-				GroupVersionKind:  gvk.VirtualService,
-				Name:              name,
-				Annotations:       routeMeta(obj),
-				Namespace:         obj.Namespace,
-				Domain:            ctx.Domain,
-			},
-			Spec: &istio.VirtualService{
-				Hosts:    []string{host},
-				Gateways: gatewayNames,
-				Tls:      routes,
-			},
-		}
-		configs = append(configs, vsConfig)
+	dest, backendErr, err := buildTCPDestination(ctx, r.BackendRefs, obj.Namespace, enforceRefGrant)
+	if err != nil {
+		return nil, err
+	}
+	return &istio.TCPRoute{
+		Route: dest,
+	}, backendErr
+}
+
+func convertTLSRoute(ctx configContext, r k8s.TLSRouteRule, obj config.Config, enforceRefGrant bool) (*istio.TLSRoute, *ConfigError) {
+	if tcpWeightSum(r.BackendRefs) == 0 {
+		// The spec requires us to reject connections when there are no >0 weight backends
+		// We don't have a great way to do it. TODO: add a fault injection API for TCP?
+		return &istio.TLSRoute{
+			Route: []*istio.RouteDestination{{
+				Destination: &istio.Destination{
+					Host:   "internal.cluster.local",
+					Subset: "zero-weight",
+					Port:   &istio.PortSelector{Number: 65535},
+				},
+				Weight: 0,
+			}},
+		}, nil
 	}
-	return configs
+	dest, backendErr, err := buildTCPDestination(ctx, r.BackendRefs, obj.Namespace, enforceRefGrant)
+	if err != nil {
+		return nil, err
+	}
+	return &istio.TLSRoute{
+		Match: buildTLSMatch(obj.Spec.(*k8s.TLSRouteSpec).Hostnames),
+		Route: dest,
+	}, backendErr
 }
 
-func buildTCPDestination(ctx configContext, forwardTo []k8s.BackendRef, ns string) ([]*istio.RouteDestination, *ConfigError) {
+func buildTCPDestination(
```

Comment:
Can you add some comments ? It's not really 'build' - but build and enforce policies, and the policies are not very obvious for some readers, and it does have pretty serious security implications. 

Maybe even split the enforcement in a separate method - but I think comments are sufficient. 

Pretty much what we discussed - if a route has parentRef as a Service, it will allow sending requests to any backend in the mesh - while if it is a Gateway it will require the destination to allow it explicitly.

Also some notes on the difference between consumer and producer routes, a link to the spec and the differences between Waypoints, Sidecars and Gateways in applying this ( it took me some time to understand - and not sure I got it right).